### PR TITLE
serial_link watchdog

### PR DESCRIPTION
**Pull request by mfine**
Take a parameter on the command line to configure a watchdog timer: if a heartbeat is not heard in `X` seconds, exit the process. Useful to drive long running health tests. Could eventually have a deeper health check here besides just simple heartbeats. Potentially there's scope for that in the new client with various configurable "watchdogs"?

/cc @fnoble @mookerji

### CODE REVIEW

**Comment by fnoble** [comment]
lgtm

File: scripts/serial_link.py [hunk 1]
```diff
@@ -426,8 +453,12 @@ def format_log_json_entry(t0, item):
...
     else:
       link.add_global_callback(default_log_callback(log_file))
-    if args.reset:
-      link.send_message(ids.RESET, '')
+  if args.reset:
+    link.send_message(ids.RESET, '')
+  # Setup watchdog
+  watchdog = args.watchdog[0]
+  if watchdog:
+    link.add_callback(0xFFFF, Watchdog(float(watchdog), default_watchdog))
```
Comment (mookerji) [inline scripts/serial_link.py]: nit: magic value: `0xFFFF` => `ids.HEARTBEAT` (I think).

File: scripts/serial_link.py [hunk 2]
```diff
@@ -382,6 +382,30 @@ def format_log_json_entry(t0, item):
...
+      self.timer.cancel()
+    self.timer = threading.Timer(self.timeout, self.alarm)
+    self.timer.start()
+
+  def alarm(self):
+    import thread
+    thread.interrupt_main()
+
+def default_watchdog():
+  print "ERROR: Watchdog expired!"
```
Comment (mookerji) [inline scripts/serial_link.py]: nit: may want to use stderr instead: `import os; os.stderr.write('ERROR: Watchdog expired!`).

File: scripts/serial_link.py [hunk 3]
```diff
@@ -382,6 +382,30 @@ def format_log_json_entry(t0, item):
   data = item.to_json_dict()
   return {'delta': delta, 'timestamp': timestamp, 'data': data}
 
+class Watchdog:
+  def __init__(self, timeout, alarm):
```
Comment (mookerji) [inline scripts/serial_link.py]: nit: a brief description, possibly saying that it calls an alarm raising a `KeyboardInterruptException` (?) or something after timeout.

File: scripts/serial_link.py [hunk 4]
```diff
@@ -382,6 +382,30 @@ def format_log_json_entry(t0, item):
...
+    self.alarm = alarm
+    self.timer = None
+
+  def __call__(self, *args):
+    self.call()
+
+  def call(self):
+    if self.timer:
+      self.timer.cancel()
+    self.timer = threading.Timer(self.timeout, self.alarm)
```
Comment (mookerji) [inline scripts/serial_link.py]: Also, lame question, but does `self.timer` need to be on a daemon thread (i.e., `self.timer.daemon = True`?

File: scripts/serial_link.py (hunk 1 above)
Comment (mfine) [inline scripts/serial_link.py]: Ha, yeah had tried that, didn't work. Looked closed, and have to do this to bring it in:

```python
from sbp_messages import SBP_HEARTBEAT
```

Isn't the wildcard in `sbp_piksi.py` doing that? Python, how does it work

Done.

File: scripts/serial_link.py (hunk 2 above)
Comment (mfine) [inline scripts/serial_link.py]: Used `sys.stderr`, since there's already some usage for `sys.stdout`. Done.

File: scripts/serial_link.py (hunk 3 above)
Comment (mfine) [inline scripts/serial_link.py]: Done.

File: scripts/serial_link.py (hunk 4 above)
Comment (mfine) [inline scripts/serial_link.py]: Totally - forgot to set this, thanks for the catch! Done.

**Comment by fnoble** [comment]
@mookerji are you happy to merge this now?

**Pull request merged**